aws/data_source_aws_dedicated_host.go

```diff
@@ -0,0 +1,104 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceAwsDedicatedHost() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsAwsDedicatedHostRead,
+
+		Schema: map[string]*schema.Schema{
+
+			"availability_zone": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"instance_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"host_recovery": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"auto_placement": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceAwsAwsDedicatedHostRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	filters, filtersOk := d.GetOk("filter")
+
+	params := &ec2.DescribeHostsInput{}
+	if filtersOk {
+		params.Filter = buildAwsDataSourceFilters(filters.(*schema.Set))
+	}
+	// var hostIDs []string
+	resp, err := conn.DescribeHosts(params)
+	if err != nil {
+		return err
+	}
+	// If no hosts were returned, return
+	if len(resp.Hosts) == 0 {
+		return fmt.Errorf("Your query returned no results. Please change your search criteria and try again.")
+	}
+
+	var filteredHosts []*ec2.Host
+
+	// loop through reservations, and remove terminated hosts, populate host slice
+	for _, host := range resp.Hosts {
+
+		if host.State != nil && *host.State != "terminated" {
```

Comment:
Can `host` be `nil` here? I don't have an IDE to check this but from squinting and looking at the value assignments, it looks like `host` is of type `*ec2.Host`. Wondering if it's useful to guard against a nil pointer or if `conn.DescribeHosts` is guaranteed to return a slice of non-nil values only.

Reply:
I think that's what already 
```if len(resp.Hosts) == 0 {
		return errors.New("Your query returned no results. Please change your search criteria and try again.")
	}```
protects against, unless we need more specific declaration for nil.

Reply:
`len(resp.Hosts) == 0` checks to see if the list is empty but not if an element of the list is nil (since it's a list of pointers). It would be very unusual so I wouldn't expect it to happen but also nil pointer access crashes your program so it's sad pandas if it ever does.